### PR TITLE
Refactor updateAssignee to use JSON-RPC

### PR DESCRIPTION
**Pull request by rsasov**

### CODE REVIEW

File: tcms/static/js/testrun_actions.js [hunk 1]
```diff
@@ -544,17 +544,9 @@ function changeCaseRunAssignee() {
     return false;
   }
 
-  jQ.ajax({
-    'url': '/runs/update-assignee/',
-    'type': 'POST',
-    'data': { ids: runs, assignee: p },
-    'success': function (data, textStatus, jqXHR) {
-      window.location.reload();
```
Comment (asankov) [inline tcms/static/js/testrun_actions.js:552]: you have removed this, and this will change behavior. we must find a way to replicate this

Comment (asankov) [reply tcms/static/js/testrun_actions.js:552]: the way I see it there are 2 options.
option 1:
```
runs.forEach(runId => jsonRPC(....., sync=true))
window.location.reload()
```
this will wait for the runs to be updated and then refresh the page. drawback - it will be if you have many runs

option 2:
```
runs.forEach(runId => jsonRPC(....., sync=false))
window.location.reload()
```
this will dispatch the calls to the backend and refresh the page without waiting for them to be completed. this will be faster, but I'm not sure how it will behave when there are many runs.

Comment (atodorov) [reply tcms/static/js/testrun_actions.js:552]: Option 2 will not work reliably. Whenever the page refreshes some requests don't seem to get through, I don't really know why. Bulk operations should always use sync=true.

File: tcms/static/js/testrun_actions.js [hunk 2]
```diff
@@ -544,17 +544,9 @@ function changeCaseRunAssignee() {
...
-    },
-    'error': function (jqXHR, textStatus, errorThrown) {
-      json_failure(jqXHR);
-    }
-  });
+  runs.forEach(function(run_id) {
+        jsonRPC('TestRun.update', [run_id, p], function(result){}, true);
+    });
```
[the remark this answers is not in the document]
Comment (atodorov) [reply tcms/static/js/testrun_actions.js:549]: Style choice is not a deal breaker ATM but needs to be consistent with everything else in that function.

File: tcms/static/js/testrun_actions.js [hunk 3]
```diff
@@ -527,28 +527,18 @@ function removeRunCC(run_id, user, container) {
...
-
-  jQ.ajax({
-    'url': '/runs/update-assignee/',
-    'type': 'POST',
-    'data': { ids: runs, assignee: p },
-    'success': function (data, textStatus, jqXHR) {
-      window.location.reload();
-    },
-    'error': function (jqXHR, textStatus, errorThrown) {
-      json_failure(jqXHR);
```
Comment (asankov) [inline tcms/static/js/testrun_actions.js:549]: FYI this is the method that shows the error alert I've posted in my other comment

**Comment by asankov** [comment]
@atodorov after offline discussion I and @rsasov think that in the end the implementation need to be something close to this:
```javascript
errors = []
testExecutions.forEach(testExecution => {
    jsonRPC('TestEx.update', ...., ()=>{}, true, error => errors.push(error))
})
if (errors) {
    // aggregate all errors into a useful error message
    json_failure(errors)
} else {
    window.location.reload()
}
```
this will be possible after `jsonRPC` is extended to receive `errorCallback` as an additional param

**Comment by asankov** [comment]
another this I noticed: when the API returns an error, we don't go to the `error`, but still go to the success one. The response body is:
```json
{"id": "jsonrpc", "jsonrpc": "2.0", "error": {"code": -32603, "message": "Internal error: [('assignee', 'Unknown user: \"something@something.com\"')]"}}
```
however, the current `jsonRPC` implementation assumes that the response has `result` property, which in this case it doesnt:
```
success: function (result) {
            callback(result.result);
      },
```
so at the point that we are inside `callback` the error is already lost.

**Comment by atodorov** [comment]
Ignore the error condition, there is a separate issue to fix that.

**Comment by asankov** [comment]
rebase to fix conflict and I think we are ready after that

File: tcms/rpc/tests/test_testexecution.py [hunk 4]
```diff
@@ -296,8 +296,8 @@ def test_update_with_non_existing_build(self):
             self.rpc_client.TestExecution.update(self.case_run_1.pk, {"build": 1111111})
 
     def test_update_with_non_existing_assignee(self):
-        with self.assertRaisesRegex(XmlRPCFault, 'Select a valid choice'):
-            self.rpc_client.TestExecution.update(self.case_run_1.pk, {"assignee": 1111111})
+        with self.assertRaisesRegex(XmlRPCFault, 'Unknown user_id'):
+            self.rpc_client.exec.TestExecution.update(self.case_run_1.pk, {"assignee": 1111111})
```
Comment (asankov) [inline tcms/rpc/tests/test_testexecution.py:300]: revert this
```suggestion
            self.rpc_client.TestExecution.update(self.case_run_1.pk, {"assignee": 1111111})
```

File: tcms/rpc/tests/test_testexecution.py [hunk 5]
```diff
@@ -296,7 +296,7 @@ def test_update_with_non_existing_build(self):
             self.rpc_client.TestExecution.update(self.case_run_1.pk, {"build": 1111111})
 
     def test_update_with_non_existing_assignee(self):
-        with self.assertRaisesRegex(XmlRPCFault, 'Select a valid choice'):
+        with self.assertRaisesRegex(XmlRPCFault, 'Unknown user_id'):
             self.rpc_client.TestExecution.update(self.case_run_1.pk, {"assignee": 1111111})
```
Comment (atodorov) [inline tcms/rpc/tests/test_testexecution.py:300]: You also need tests for non-existing username/email, same thing for existing username/email/id because all of the 3 are possible values.

File: tcms/rpc/tests/test_testexecution.py [hunk 6]
```diff
@@ -291,13 +291,45 @@ def test_update_with_single_caserun(self):
         self.assertEqual(tcr['status'], self.status_positive.name)
         self.assertEqual(tcr['sortkey'], 90)
 
+    def test_update_with_assignee_id(self):
+        tcr = self.rpc_client.TestExecution.update(self.case_run_1.pk, {
```
Comment (atodorov) [inline tcms/rpc/tests/test_testexecution.py:295]: `tcr` -> `result` or `execution`. tcr doesn't have any useful meaning nowadays.

File: tcms/rpc/tests/test_testexecution.py [hunk 7]
```diff
@@ -291,13 +291,45 @@ def test_update_with_single_caserun(self):
...
+            "assignee": self.user.pk
+        })
+        self.assertEqual(tcr['assignee'], self.user.username)
+
+    def test_update_with_assignee_email(self):
+        tcr = self.rpc_client.TestExecution.update(self.case_run_1.pk, {
```
Comment (atodorov) [inline tcms/rpc/tests/test_testexecution.py:301]: Before manipulating the assignee field tests should ensure that it is different from the target value. A simple assertNotEqual will do.

Note: technically not necessary b/c TestExecutionFactory will create a new user if not provided, but we don't want to rely blindly on this and it's better to explicitly check preconditions.

File: tcms/rpc/tests/test_testexecution.py [hunk 8]
```diff
@@ -58,20 +58,20 @@ def test_create_with_no_required_fields(self):
                 self.rpc_client.TestExecution.create(value)
 
     def test_create_with_required_fields(self):
-        tcr = self.rpc_client.TestExecution.create({
+        execution = self.rpc_client.TestExecution.create({
```
Comment (atodorov) [inline tcms/rpc/tests/test_testexecution.py:61]: Next time when you want to rename something outside the main scope do so in a separate commit.

Previous code review talks about renaming variables only in the new tests.
